eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxosSpec.hs

```diff
@@ -0,0 +1,201 @@
+  , PredicateFailure (EraRule "UTXO" era) ~ BabbageUtxoPredFailure era
+  , PredicateFailure (EraRule "UTXOS" era) ~ AlonzoUtxosPredFailure era
+  , ContextError era ~ ConwayContextError era
+  ) =>
+  SpecWith (ImpTestState era)
+spec =
+  describe "UTXOS" $ do
+    it "can use reference scripts" $ do
+      producingTx <- setupRefTx
+      referringTx <-
+        submitTxAnn "Transaction that refers to the script" $
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . inputsTxBodyL .~ Set.singleton (mkTxInPartial producingTx 1)
+            & bodyTxL . referenceInputsTxBodyL .~ Set.singleton (mkTxInPartial producingTx 0)
+      (referringTx ^. witsTxL . scriptTxWitsL) `shouldBe` mempty
+    it "can use regular inputs for reference" $ do
+      producingTx <- setupRefTx
+      referringTx <-
+        submitTxAnn "Consuming transaction" $
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . inputsTxBodyL
+              .~ Set.fromList
+                [ mkTxInPartial producingTx 0
+                , mkTxInPartial producingTx 1
+                ]
+      (referringTx ^. witsTxL . scriptTxWitsL) `shouldBe` mempty
+    it "fails with same txIn in regular inputs and reference inputs" $ do
+      producingTx <- setupRefTx
+      let
+        consumingTx =
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . inputsTxBodyL
+              .~ Set.fromList
+                [ mkTxInPartial producingTx 0
+                , mkTxInPartial producingTx 1
+                ]
+            & bodyTxL . referenceInputsTxBodyL .~ Set.singleton (mkTxInPartial producingTx 0)
+      _ <-
+        submitFailingTx
+          consumingTx
+          [ ConwayUtxowFailure . UtxoFailure . BabbageNonDisjointRefInputs @era $
+              mkTxInPartial producingTx 0 :| []
+          ]
+      pure ()
+    it "fails when using inline datums for PlutusV1" $ do
+      shSpending <- setupPlutusV1Script
+      refTxOut <- mkRefTxOut shSpending
+      producingTx <-
+        fmap txIdTx . submitTxAnn "Producing transaction" $
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . outputsTxBodyL
+              .~ SSeq.fromList
+                [ refTxOut
+                , scriptLockedTxOut shSpending & dataTxOutL .~ SJust (Data $ P1.I 3)
+                ]
+      let
+        lockedTxIn = mkTxInPartial producingTx 1
+        consumingTx =
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . inputsTxBodyL .~ Set.singleton lockedTxIn
+            & bodyTxL . referenceInputsTxBodyL .~ Set.singleton (mkTxInPartial producingTx 0)
+      impAnn "Consuming transaction" $
+        submitFailingTx
+          consumingTx
+          [ ConwayUtxowFailure
+              . UtxoFailure
+              . AlonzoInBabbageUtxoPredFailure
+              . UtxosFailure
+              $ CollectErrors
+                [ BadTranslation
+                    . BabbageContextError
+                    . InlineDatumsNotSupported
+                    $ TxOutFromInput lockedTxIn
```

Comment:
This will allow usage of this test in future eras
```suggestion
                [ BadTranslation
                    . inject
                    $ TxOutFromInput lockedTxIn
```